rust/db_handling/src/identities.rs

```diff
@@ -126,6 +133,90 @@ pub fn generate_random_phrase(words_number: u32) -> Result<String> {
     Ok(mnemonic.into_phrase())
 }
 
+/// Check that key with a given path should be marked as a progeny of a key with
+/// exposed secret.
+///
+/// There could be false positives here, as the passwords are not stored
+/// anywhere in Signer, therefore if there are two passwords, there is no way to
+/// check if they are different.
+///
+/// <table>
+///     <tr><th>current</th><th>mark current?</th><th>exposed in set</th></tr>
+///     <tr><td><code>//Alice</code></td><td>-</td><td><code>//Ali</code></td></tr>
+///     <tr><td><code>//Alice//1</code></td><td>+</td><td><code>//Alice</code></td></tr>
+///     <tr><td><code>//Alice///&ltpassword&gt</code></td><td>+</td><td><code>//Alice</code></td></tr>
+///     <tr><td><code>//Alice//1///&ltpassword&gt</code></td><td>+</td><td><code>//Alice</code></td></tr>
+///     <tr><td><code>//Alice</code></td><td>-</td><td><code>//Alice///&ltpassword&gt</code></td></tr>
+///     <tr><td><code>//Alice//1</code></td><td>-</td><td><code>//Alice///&ltpassword&gt</code></td></tr>
+///     <tr><td><code>//Alice///&ltpassword1&gt</code></td><td>+</td><td><code>//Alice///&ltpassword0&gt</code></td></tr>
+///     <tr><td><code>//Alice//1///&ltpassword1&gt</code></td><td>+</td><td><code>//Alice///&ltpassword0&gt</code></td></tr>
+///
+/// </table>
+#[cfg(any(feature = "active", feature = "signer"))]
+pub(crate) fn is_potentially_exposed(
+    path: &str,
+    path_is_passworded: bool,
+    exposed_path: &str,
+    exposed_is_passworded: bool,
+) -> bool {
+    if (path == exposed_path) || path.starts_with(&format!("{}/", exposed_path)) {
+        path_is_passworded || !exposed_is_passworded
+    } else {
+        false
+    }
+}
+
+/// Check if given key should be marked as exposed because its parent was
+/// exposed.
+///
+/// Input set is already filtered by seed name elsewhere.
+#[cfg(any(feature = "active", feature = "signer"))]
+fn has_parent_with_exposed_secret(
+    new_cropped_path: &str,
+    new_is_passworded: bool,
+    seed_name: &str,
+    database_name: &str,
+) -> Result<bool> {
+    let mut out = false;
+    for (_, address_details) in get_addresses_by_seed_name(database_name, seed_name)?.iter() {
+        if address_details.secret_exposed
+            && is_potentially_exposed(
+                new_cropped_path,
+                new_is_passworded,
+                &address_details.path,
+                address_details.has_pwd,
+            )
+        {
+            out = true;
+            break;
+        }
+    }
+    Ok(out)
```

Comment:
This can be simplified a bit: 

```suggestion
    get_addresses_by_seed_name(database_name, seed_name)?.iter().any(|(_, address_details)| {
        address_details.secret_exposed
            && is_potentially_exposed(
                new_cropped_path,
                new_is_passworded,
                &address_details.path,
                address_details.has_pwd,
            )
         }
     }
```